File: python/export_result.py
"""
Hydrophone Deconvolution - Multi-language Implementation

Based on Tutorial-Deconvolution by Weber & Wilkens (2023)
DOI: 10.5281/zenodo.10079801
License: CC BY 4.0
"""

import numpy as np
from pathlib import Path
from deconvolution.core import deconvolve_without_uncertainty, pulse_parameters


def main():
    data_dir = Path(__file__).parent.parent / "test-data"
    results_dir = Path(__file__).parent.parent / "validation" / "results"
    results_dir.mkdir(parents=True, exist_ok=True)

    measured_signal = np.loadtxt(data_dir / "measured_signal.csv")
    freq_real = np.loadtxt(data_dir / "freq_response_real.csv")
    freq_imag = np.loadtxt(data_dir / "freq_response_imag.csv")
    frequency_response = freq_real + 1j * freq_imag

    result = deconvolve_without_uncertainty(measured_signal, frequency_response, 1e7)
    np.savetxt(results_dir / "python_result.csv", result, fmt="%.18e")
    print(f"Python result exported: {len(result)} samples")

    # Pulse parameters
    signal_uncertainty = np.loadtxt(data_dir / "signal_uncertainty.csv")
    n = len(result)
    sampling_rate = 1e7
    time = np.arange(n) / sampling_rate
    pp = pulse_parameters(time, result, signal_uncertainty)
    with open(results_dir / "python_pulse_params.csv", "w") as f:
        f.write(f"{pp['pc_value']:.18e},{pp['pc_uncertainty']:.18e},"
                f"{pp['pr_value']:.18e},{pp['pr_uncertainty']:.18e},"
                f"{pp['ppsi_value']:.18e},{pp['ppsi_uncertainty']:.18e}\n")
    print(f"Python pulse parameters exported")


if __name__ == "__main__":
    main()

File: rust/src/bin/export_result.rs
//! Hydrophone Deconvolution - Multi-language Implementation
//!
//! Based on Tutorial-Deconvolution by Weber & Wilkens (2023)
//! DOI: 10.5281/zenodo.10079801
//! License: CC BY 4.0

use hydrophone_deconvolution::{deconvolve_without_uncertainty, pulse_parameters, PulseUncertainty};
use num_complex::Complex64;
use std::fs;
use std::io::Write;

fn load_csv(path: &str) -> Vec<f64> {
    fs::read_to_string(path)
        .unwrap()
        .lines()
        .filter(|l| !l.trim().is_empty())
        .map(|l| l.trim().parse::<f64>().unwrap())
        .collect()
}

fn main() {
    let data_dir = "/workspace/test-data";
    let results_dir = "/workspace/validation/results";
    fs::create_dir_all(results_dir).unwrap();

    let measured = load_csv(&format!("{}/measured_signal.csv", data_dir));
    let freq_real = load_csv(&format!("{}/freq_response_real.csv", data_dir));
    let freq_imag = load_csv(&format!("{}/freq_response_imag.csv", data_dir));

    let freq_response: Vec<Complex64> = freq_real
        .iter()
        .zip(freq_imag.iter())
        .map(|(&r, &i)| Complex64::new(r, i))
        .collect();

    let result = deconvolve_without_uncertainty(&measured, &freq_response, 1e7);

    let mut file = fs::File::create(format!("{}/rust_result.csv", results_dir)).unwrap();
    for val in &result {
        writeln!(file, "{:.18e}", val).unwrap();
    }

    println!("Rust result exported: {} samples", result.len());

    // Pulse parameters
    let sig_unc = load_csv(&format!("{}/signal_uncertainty.csv", data_dir));
    let n = result.len();
    let sampling_rate = 1e7;
    let time_vec: Vec<f64> = (0..n).map(|i| i as f64 / sampling_rate).collect();

    let pp = pulse_parameters(&time_vec, &result, PulseUncertainty::Vector(sig_unc));

    let mut pp_file = fs::File::create(format!("{}/rust_pulse_params.csv", results_dir)).unwrap();
    writeln!(pp_file, "{:.18e},{:.18e},{:.18e},{:.18e},{:.18e},{:.18e}",
        pp.pc_value, pp.pc_uncertainty, pp.pr_value, pp.pr_uncertainty,
        pp.ppsi_value, pp.ppsi_uncertainty).unwrap();

    println!("Rust pulse parameters exported");
}
